### PR TITLE
Investigation of Phase-Error-Propagation through a FODO lattice in cpymad

### PR DESCRIPTION
**Pull request by JoschD**
A small study aiming to investigate how phase errors, 
determined by the errors on alpha and beta of the boundary conditions,
propagate through the lattice.

In particular, this is a test whether the first plus-sign in 
Eq. (2) of 10.18429/JACOW-IPAC2015-MOPJE054  
( http://jacow.org/ipac2015/doi/JACoW-IPAC2015-MOPJE054.html )
should in fact be a minus sign.

### CODE REVIEW

File: FODO_Test_Lattice/Phase_Error_Propagation/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+determined by the errors on alpha and beta of the boundary conditions,
+propagate through the lattice.
+
+In particular, this is a test whether the first plus-sign in 
+Eq. (2) of 10.18429/JACOW-IPAC2015-MOPJE054  
+( http://jacow.org/ipac2015/doi/JACoW-IPAC2015-MOPJE054.html )
+should in fact be a minus sign:
+Suitable alpha0 and beta0, means and standard deviations, 
+are choosen and realized as boundary conditions for the FODO-lattice from 
+gaussian distributions.
```
Comment (mihofer) [inline FODO_Test_Lattice/Phase_Error_Propagation/README.md:15]: Gaussian

File: FODO_Test_Lattice/Phase_Error_Propagation/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Phase Error Propagation
+
+## Motivation 
+
+This small study aims to investigate how phase errors, 
+determined by the errors on alpha and beta of the boundary conditions,
+propagate through the lattice.
```
Comment (mihofer) [inline FODO_Test_Lattice/Phase_Error_Propagation/README.md:7]: not sure this is easily understandable .. 
rather saying "this small study aims to investigate how phase errors propagate through the lattice, 
using a given alpha and beta including errors as boundary condition."

File: FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    beta, errbeta = init.as_tuple(init.beta)
+
+    res = np.sqrt(
+        (
+                (((np.cos(4 * np.pi * dphi) - 1) * alpha) - 
+                 (np.sin(4 * np.pi * dphi))
+                 ) * 0.5 * errbeta/beta
+        ) ** 2 +
+        ((np.cos(4*np.pi*dphi) - 1) * 0.5 * erralpha) ** 2
+    ) / (2*np.pi)
```
Comment (mihofer) [inline FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py:175]: making this formula a fct with input of sign avoids copy-paste errors and maybe better readable

File: FODO_Test_Lattice/Phase_Error_Propagation/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+are choosen and realized as boundary conditions for the FODO-lattice from 
+gaussian distributions.
+From the resulting phase advance distributions, 
+shifted to zero by the phase advance of the realization with only mean-values, 
+the standard deviation is calculated and compared to
+the analytically propagated results.
+
+The resulting plots show that, indeed, the error propagation formular with
+a negative sign is the correct one.
+![Result for 1000 realizations](result_1000realizations.png "Result for 1000 realizations")
```
Comment (mihofer) [inline FODO_Test_Lattice/Phase_Error_Propagation/README.md:23]: bit hard to infer what and why you come to the conclusion ..
and additional normal distribution as overflay with sigmas marked or so could help

File: FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+# Helper ---
+
+def generate_fodo(sequence: str ="fodo") -> Madx:
+    """Generate a standard FODO sequence.
+    
+    Args:
+        sequence (str): Sequence name.
+    """
+    madx = Madx()
```
Comment (mihofer) [inline FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py:119]: for general (re)use, wouldn't it be better to provide the cpyMAD-X instance as input to the fct?

Comment (mihofer) [reply FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py:119]: and for more fun studies, the phase adv. of the cell could also be an input, to compare e.g. 60 and 90 deg LHC stuff

File: FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    The mean and std values are hardcoded and choosen to have a final phase advance 
+    of around 0.25 (far away from 0.5) in units of 2 Pi.
+
+    Args:
+        output (Path): Path to save the output tfs-file.
+        n (int, optional): Number of realizations to generate. Defaults to 100.
+    """
+    madx = generate_fodo()
+    madx.twiss(chrom=True)
+
```
Comment (mihofer) [inline FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py:67]: add comment # define boundary conditions with uncertainty using uncertainties.ufloat(value, err) or similar

File: FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py (hunk 3 above)
Comment (mihofer) [reply FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py:175]: if you have the inputs as ufloats, can't you also use the direct formula phi= ... and let uncertainties do the math to then get the err on phi afterwards?

File: FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py (hunk 6 above)
Comment (JoschD) [reply FODO_Test_Lattice/Phase_Error_Propagation/phase_propagation.py:67]: I would say that is pretty clear from the class-name, also given the doc-string of the function.

File: FODO_Test_Lattice/Phase_Error_Propagation/README.md (hunk 4 above)
Comment (JoschD) [reply FODO_Test_Lattice/Phase_Error_Propagation/README.md:23]: added comment about the STD is represented by the green dashed line

**Review by mihofer — APPROVED** [review]
